src/libstd/sys/cloudabi/mutex.rs

```diff
@@ -47,25 +48,28 @@ impl Mutex {
 }
 
 pub struct ReentrantMutex {
-    lock: UnsafeCell<AtomicU32>,
-    recursion: UnsafeCell<u32>,
+    lock: UnsafeCell<MaybeUninit<AtomicU32>>,
+    recursion: UnsafeCell<MaybeUninit<u32>>,
 }
 
 impl ReentrantMutex {
     pub unsafe fn uninitialized() -> ReentrantMutex {
-        mem::uninitialized()
+        ReentrantMutex { 
+            lock: UnsafeCell::new(MaybeUninit::uninit()),
+            recursion: UnsafeCell::new(MaybeUninit::uninit())
+        }
     }
 
     pub unsafe fn init(&mut self) {
-        self.lock = UnsafeCell::new(AtomicU32::new(abi::LOCK_UNLOCKED.0));
-        self.recursion = UnsafeCell::new(0);
+        self.lock = UnsafeCell::new(MaybeUninit::new(AtomicU32::new(abi::LOCK_UNLOCKED.0)));
+        self.recursion = UnsafeCell::new(MaybeUninit::new(0));
     }
 
     pub unsafe fn try_lock(&self) -> bool {
         // Attempt to acquire the lock.
         let lock = self.lock.get();
         let recursion = self.recursion.get();
-        if let Err(old) = (*lock).compare_exchange(
+        if let Err(old) = (*(*lock).as_mut_ptr()).compare_exchange(
```

Comment:
Considering that [there is no `Deref` impl for `UnsafeCell`](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=218f654f5104391a71e8b605bdf1e926), how is this supposed to compile?

Reply:
It compiles because in the lines above (70 and 71) `lock` and `recursion` are bound to mut pointers to the `UnsafeCell` contents, so it's just a deref of the pointer.

Reply:
Ah, that makes so much more sense. :)

Reply:
Would it make sense to call `as_mut_ptr` up there to avoid having to all it everywhere inside the function?